### PR TITLE
Add delete endpoint for applications

### LINKED ISSUE #310: Create REST endpoint for deletion of MicoApplications
As a user I want to have the possibility to delete existing MicoApplications.

### PR DESCRIPTION
**Pull request by st116290**
adds the missing delete endpoint `/applications/{shortName}/{version}`

closes #310

### CODE REVIEW

**Comment by st116290** [comment]
This is a reimplementation of #208 with the new domain model.

File: mico-core/src/main/java/io/github/ust/mico/core/REST/ApplicationController.java [hunk 1]
```diff
@@ -155,6 +156,31 @@
...
+    public ResponseEntity<Resource<MicoApplication>> deleteApplication(@PathVariable(PATH_VARIABLE_SHORT_NAME) String shortName,
+                                                                       @PathVariable(PATH_VARIABLE_VERSION) String version) {
+        Optional<MicoApplication> applicationOptional = applicationRepository.findByShortNameAndVersion(shortName, version);
+        if (!applicationOptional.isPresent()) {
+            // application already deleted
+            return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+        }
+        applicationOptional.map(application -> {
+            if (application.getDeploymentInfo() == null || application.getDeploymentInfo().getServiceDeploymentInfos() != null) {
+                return application; // TODO better deployment detection
```
Comment (JanVetter) [inline mico-core/src/main/java/io/github/ust/mico/core/REST/ApplicationController.java]: Do we really want to return the deleted application?

File: mico-core/src/main/java/io/github/ust/mico/core/REST/ApplicationController.java [hunk 2]
```diff
@@ -155,6 +156,31 @@
...
+    @DeleteMapping("/{" + PATH_VARIABLE_SHORT_NAME + "}/{" + PATH_VARIABLE_VERSION + "}")
+    public ResponseEntity<Resource<MicoApplication>> deleteApplication(@PathVariable(PATH_VARIABLE_SHORT_NAME) String shortName,
+                                                                       @PathVariable(PATH_VARIABLE_VERSION) String version) {
+        Optional<MicoApplication> applicationOptional = applicationRepository.findByShortNameAndVersion(shortName, version);
+        if (!applicationOptional.isPresent()) {
+            // application already deleted
+            return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+        }
+        applicationOptional.map(application -> {
+            if (application.getDeploymentInfo() == null || application.getDeploymentInfo().getServiceDeploymentInfos() != null) {
```
Comment (JanVetter) [inline mico-core/src/main/java/io/github/ust/mico/core/REST/ApplicationController.java]: I don't really get this.. Isn't this somehow double-checking?

File: mico-core/src/main/java/io/github/ust/mico/core/REST/ApplicationController.java [hunk 3]
```diff
@@ -155,6 +156,31 @@
...
+        }).map(application -> {
+            return application.toBuilder().clearServices().build();
+        }).map(application -> {
+            applicationRepository.save(application);
+            return application;
+        }).map(application -> {
+            applicationRepository.delete(application);
+            return application;
+        });
+        return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
```
Comment (JanVetter) [inline mico-core/src/main/java/io/github/ust/mico/core/REST/ApplicationController.java]: Why do we return `HttpStatus.NO_CONTENT`? Wouldn't it be better so return some `SUCCESS`?

File: mico-core/src/test/java/io/github/ust/mico/core/ApplicationControllerTest.java [hunk 4]
```diff
@@ -215,6 +219,23 @@ public void updateApplication() throws Exception {
         resultUpdate.andExpect(status().isOk());
     }
 
+    @Test
+    public void deleteApplication() throws Exception {
+        MicoApplication app = MicoApplication.builder().shortName(SHORT_NAME).version(VERSION).build();
+        given(applicationRepository.findByShortNameAndVersion(SHORT_NAME, VERSION)).willReturn(Optional.of(app));
+
+        ArgumentCaptor<MicoApplication> appCaptor = ArgumentCaptor.forClass(MicoApplication.class);
```
Comment (JanVetter) [inline mico-core/src/test/java/io/github/ust/mico/core/ApplicationControllerTest.java]: What is this?

**Pull request merged**